src/main/java/graphql/schema/GraphQLAppliedDirectiveArgument.java

```diff
@@ -0,0 +1,252 @@
+package graphql.schema;
+
+
+import graphql.PublicApi;
+import graphql.language.Argument;
+import graphql.language.Value;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static graphql.Assert.assertValidName;
+import static graphql.execution.ValuesResolver.getInputValueImpl;
+
+/**
+ * This defines an argument that can be supplied to a graphql field (via {@link GraphQLFieldDefinition}.
+ * <p>
+ * Fields can be thought of as "functions" that take arguments and return a value.
+ * <p>
+ * See http://graphql.org/learn/queries/#arguments for more details on the concept.
+ * <p>
+ * {@link GraphQLAppliedDirectiveArgument} is used in two contexts, one context is graphql queries where it represents the arguments that can be
+ * set on a field and the other is in Schema Definition Language (SDL) where it can be used to represent the argument value instances
+ * that have been supplied on a {@link GraphQLDirective}.
+ * <p>
+ * The difference is the 'value' and 'defaultValue' properties.  In a query argument, the 'value' is never in the GraphQLArgument
+ * object but rather in the AST direct or in the query variables map and the 'defaultValue' represents a value to use if both of these are
+ * not present. You can think of them like a descriptor of what shape an argument might have.
+ * <p>
+ * However with directives on SDL elements, the value is specified in AST only and transferred into the GraphQLArgument object and the
+ * 'defaultValue' comes instead from the directive definition elsewhere in the SDL.  You can think of them as 'instances' of arguments, their shape and their
+ * specific value on that directive.
+ */
+@PublicApi
+public class GraphQLAppliedDirectiveArgument implements GraphQLNamedSchemaElement {
```

Comment:
regarding naming: so the question is: has this actually anything todo with directives? How about `GraphQLArgumentValue`?

Reply:
In some ways they dont have anything to do with directives - HOWEVER the only place where we have a "name" and value is in directives in SDL and "args" in queries

We never created a `GraphQLArgumentValue` in queries (just map<string,object> but yeah we could.  

So I like `GraphQLArgumentValue ` or maybe `GraphqlAppliedArgument`

Reply:
renamed